keps/78-dynamically-reclaiming-resources/README.md

```diff
@@ -0,0 +1,142 @@
+
+Whenever a Pod of a Job fails, the Job will recreate a Pod as replacement against the failed Pod. The Job will continue this process of creating new Pods until the Job reaches its `backoffLimit`(by default the backoffLimit is `6`). The newly created Pod against the failed Pod will reuse the resources of the failed Pod. Once the Job reaches its `backoffLimit` and does not have the required successful `.spec.completions` count, the Job is termed as a `Failed` Job. When the Job is marked as failed, no more Pods would be created by the Job. So, the remaining owned resources of the Job will be reclaimed by queue. 
+
+#### RestartPolicy of Pods as OnFailure
+
+When the Pods of the Job have `.spec.template.spec.restartPolicy = "OnFailure"`, the failed Pod stays on the node and the failed container within the Pod is restarted. The Pod might run indefinitely or get marked as `Failed` after the `.spec.activeDeadlineSeconds` is completed by Job if specified. In this case also, Kueue won't reclaim the resources of the Job until the Job gets completed as either `Failed` or `Succeeded`.
+
+Hence, as seen from the above discussed Pod failure cases, we conclude that the Pods of a Job which fail during its execution, its resources are not immediately reclaimed by Kueue. Only when the Job gets marks as `Failed`, the resources of failed Pods will be reclaimed by Kueue.
+
+## API
+
+A new field `ReclaimedPodSetPods` is added to the `.status` of Workload API.
+
+```go
+// WorkloadStatus defines the observed state of Workload
+type WorkloadStatus struct {
+    // conditions hold the latest available observations of the Workload
+    // current state.
+    // +optional
+    // +listType=map
+    // +listMapKey=type
+    Conditions []WorkloadCondition `json:"conditions,omitempty"`
+
+    // list of count of Pods of a PodSet with resources reclaimed
+    // +optional
+    ReclaimedPodSetPods []ReclaimedPodSetPods `json:"reclaimedPodSetPods"`
+}
+
+// ReclaimedPodSetPods defines the PodSet name and count of successfully completed Pods 
+// belonging to the PodSet whose resources have been reclaimed by Kueue
+type ReclaimedPodSetPods struct{
+    Name string
+    Count int
+}
+```
+
+#### ReclaimedPodSetPods
+`.status.reclaimedPodSetPods` is a list where each element of the list denotes the count of Pods of the Workload w.r.t PodSet whose resources have been reclaimed by Kueue. The structure consists of two fields - `Name` denotes the name of PodSet and `Count` denotes the count of Pods belonging to the `Name` PodSet.
+
+## Implementation
+
+The workload reconciler will keep a watch(or `Watches()`) on the Job's `.status.succeeded` field. The Workload reconciler will calculate the difference between the Job's `.status.succeeded` field and Workload's `.status.reclaimedPodSetPods[i].count` field for every PodSet of the workload. With the former value being greater than the later, Kueue will reclaim the resources of the excess succeeded Pods.
```

Comment:
The workload reconciler doesn't watch Job.

Did you mean the job reconciler?

Shouldn't `count` be updated based on `succeeded`? Although, it's unclear what to do when `parallelism != completions`

Reply:
I meant Job CR only.

Yes, will rethink for the case where `parallelism != completions`

Reply:
> The workload reconciler doesn't watch Job.

+ 1, or it will be too heavy for workload controller and coupled with Jobs.